Feature Gate - Kubeadm Audit Logging

chuckha opened the pull request:
Fixes kubernetes/kubeadm#623

Signed-off-by: Chuck Ha <ha.chuck@gmail.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
This PR enables [Auditing](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/) behind a featureGate. A user can supply their own audit policy with configuration option as well as a place for the audit logs to live. If no policy is supplied a default policy will be provided. The default policy will log all Metadata level policy logs. It is the example provided in the documentation.
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes kubernetes/kubeadm#623

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
kubeadm: Enable auditing behind a feature gate.
```

timothysc commented on cmd/kubeadm/app/phases/controlplane/manifests.go:
```diff
@@ -128,6 +129,12 @@ func createStaticPodFiles(manifestDir string, cfg *kubeadmapi.MasterConfiguratio
 	}
 
+	// create audit log file for apiserver
+	if err := audit.CreateDefaultAuditLogPolicy(cfg.AuditPolicyDir); err != nil {
```
People are going to want to knob paths here for policy on logging.  Given it's beta I'm debating whether we should feature gate it or just provide a 1st class override. 

/cc @roberthbailey @liggitt .

liggitt replied:
seems like a natural point to drop in a custom policy, especially since nothing else in kubeadm is really functionally dependent on this particular default audit policy

timothysc reviewed (changes requested):
Need some feedback on knob & default before we continue.

dixudx commented on cmd/kubeadm/app/util/audit/utils.go:
```diff
@@ -0,0 +1,71 @@
+		},
+		Rules: []auditv1beta1.PolicyRule{
+			{
+				Level: auditv1beta1.LevelMetadata,
```
Users cannot customize the rules here.

timothysc replied:
How do you propose we expose?

dixudx replied:
@chuckha @timothysc Let the users use their own policy files optionally.
Just make a slight change to current `AuditPolicyDir` behavior.
If `AuditPolicyDir` is not set, create and use this default policy file. If set, find that file named `audit.yaml` and use it.

timothysc replied:
sgtm!

dixudx commented on cmd/kubeadm/app/util/audit/utils.go:
```diff
@@ -0,0 +1,71 @@
+	// Registers auditv1beta1 with the runtime Scheme
+	auditv1beta1.AddToScheme(scheme.Scheme)
+
+	// writes the pod to disk
```
s/pod/policy

timothysc commented on cmd/kubeadm/app/apis/kubeadm/v1alpha1/defaults.go:
```diff
@@ -99,6 +105,19 @@ func SetDefaults_MasterConfiguration(obj *MasterConfiguration) {
+		obj.AuditPolicyLogDir = DefaultAuditPolicyLogDir
+	}
+
+	// TODO(chuckha) is 0 a meaningful value here? Does it mean 0 days or forever or can it be used as the empty value?
```
Need to rationalize this comment.

chuckha replied:
https://github.com/natefinch/lumberjack/blob/v2.0/lumberjack.go#L78

0 is a valid number and, in addition to MaxBackups == 0, means nothing will get deleted. I'll need to update this to accept 0 as valid and meaningful.

chuckha commented on cmd/kubeadm/app/constants/constants.go:
```diff
@@ -65,6 +65,11 @@ const (
+	// AuditPolicyFileName defines audit policy resource file for apiserver
+	AuditPolicyFileName = "audit.yaml"
+	// AuditPolicyLogFileName defines the audit log filename
+	AuditPolicyLogFileName = "audit.log"
```
Audit log uses lumberjack under the hood which uses timestamps embedded in the filename to determine age of logs. I don't think defining an AuditPolicyLogFileName is the right thing to do here, unless I'm missing something.

chuckha replied:
nope, my misundertsanding. lumberjack will take care of this automatically after the max-size.

timothysc reviewed (approved):
/approve - minor style nits, but I don't feel strongly about them.  

@dixudx - poke for once over and lgtm

liggitt commented on cmd/kubeadm/app/phases/controlplane/manifests.go:
```diff
@@ -128,6 +129,15 @@ func createStaticPodFiles(manifestDir string, cfg *kubeadmapi.MasterConfiguratio
 	}
 
+	// use the default audit policy if no audit policy exists
+	auditPolicyFilePath := filepath.Join(cfg.AuditPolicyDir, kubeadmconstants.AuditPolicyFileName)
```
Pinned file names inside user specified directories makes for a really opaque interface. Is there a reason not to simply have the user specify the policy file directly?

chuckha replied:
I could add an `--audit-policy-file` flag to combine with the `--audit-policy-dir` flag to make this customizable but default to `audit.yaml`. Does that sound reasonable?

liggitt replied:
why not just `--audit-policy-file` as the full filepath and drop `--audit-policy-dir`?

chuckha replied:
that seems sensible! 👍 thanks for the input

chuckha committed:
Enable Audit Logs Behind a Feature Gate

Audit logs are configurable via the MasterConfiguration file.

All options are ignored unless the FeatureGate is enabled.

Fixes kubernetes/kubeadm#623

Signed-off-by: Chuck Ha <ha.chuck@gmail.com>

ncdc commented:
/ok-to-test

ncdc commented:
/test all

chuckha commented on cmd/kubeadm/app/apis/kubeadm/v1alpha1/defaults.go:
```diff
@@ -211,3 +223,16 @@ func SetDefaults_KubeletConfiguration(obj *MasterConfiguration) {
+
+// SetDefaults_AuditPolicyConfiguration sets default values for the AuditPolicyConfiguration
+func SetDefaults_AuditPolicyConfiguration(obj *MasterConfiguration) {
+	if obj.AuditPolicyConfiguration.Path == "" {
```
`""` is actually the default I want here.

kargakis commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -191,6 +192,18 @@ func AddInitConfigFlags(flagSet *flag.FlagSet, cfg *kubeadmapiext.MasterConfigur
 		"The duration before the bootstrap token is automatically deleted. If set to '0', the token will never expire.",
 	)
+	flagSet.StringVar(
+		&cfg.AuditPolicyConfiguration.Path, "audit-policy", cfg.AuditPolicyConfiguration.Path,
```
I think there was agreement in yesterday's SIG meeting to not proceed with adding flags until we sort out whether we want to be supporting these in the future.

@kubernetes/sig-cluster-lifecycle-misc

timothysc replied:
I think we may have to default off with a opt-in via config option for the time being.  Due to the logistical constraints on both args, and self-hosting HA concerns.  

This will enable folks to turn on b/default with a separate PR, and unblock this one.

fabriziopandini reviewed (commented):
@chuckha :+1 for this PR!

Two high level comments about the implementation:
- if we are extending how static pod manifest are created in `kubeadm init`, it is necessary to ensure a consistent behaviour also in `kubeadm alpha phase controlplane` (e.g. same flags, same logic for the creation of the policy file if it doesn't exists). 
- the dependency from the kubea-apiserver to the audit policy file introduce a new element that should be taken into account for self-hosting and HA. I think it is important to inform as many people as possible about this. @timothysc, what about discussing this in the next SIG-meeting?

xiangpengzhao commented on cmd/kubeadm/app/apis/kubeadm/fuzzer/fuzzer.go:
```diff
@@ -111,6 +111,12 @@ func Funcs(codecs runtimeserializer.CodecFactory) []interface{} {
+			obj.AuditPolicyConfiguration = kubeadm.AuditPolicyConfiguration{
+				Path:      "foo",
+				LogDir:    "/foo",
+				LogMaxAge: &zero,
```
If the type of `LogMaxAge` is going to be changed to `int32`, then you can use `utilpointer.Int32Ptr(0)` instead :)

chuckha replied:
sounds good!

xiangpengzhao commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -259,3 +262,13 @@ type HostPathMount struct {
+	// LogDir is the local pth to the directory where logs should be stored.
+	LogDir string
+	// LogMaxAge is the number of days logs will be stored for. 0 indicates forever.
+	LogMaxAge *int
```
How about changing the type to `int32`?

xiangpengzhao commented on cmd/kubeadm/app/apis/kubeadm/v1alpha1/defaults.go:
```diff
@@ -67,6 +67,15 @@ const (
 	KubeproxyKubeConfigFileName = "/var/lib/kube-proxy/kubeconfig.conf"
+
+	// DefaultAuditPolicyLogDir is the default directory where the audit system will write
+	DefaultAuditPolicyLogDir = "/var/log/kubernetes"
```
How about creating a sub-dir? This is OK anyway.

timothysc replied:
@xiangpengzhao - Do you mean /var/log/kubernetes/audit ?

xiangpengzhao replied:
@timothysc exactly :)

xiangpengzhao commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -259,3 +262,13 @@ type HostPathMount struct {
+	LogDir string
+	// LogMaxAge is the number of days logs will be stored for. 0 indicates forever.
+	LogMaxAge *int
+}
```
Add the same TODO here? i.e., `//TODO(chuckha) add other options for audit policy`

xiangpengzhao commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -319,6 +320,21 @@ func (i *Init) Run(out io.Writer) error {
+		} else {
+			i.cfg.AuditPolicyConfiguration.Path = filepath.Join(kubeConfigDir, kubeadmconstants.AuditPolicyDir, kubeadmconstants.AuditPolicyFile)
+			if err := auditutil.CreateDefaultAuditLogPolicy(i.cfg.AuditPolicyConfiguration.Path); err != nil {
+				return fmt.Errorf("error creating default audit policy [%v]", err)
```
Also print the default file path here?

chuckha replied:
sure, seems reasonable

xiangpengzhao commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -259,3 +262,13 @@ type HostPathMount struct {
+type AuditPolicyConfiguration struct {
+	// Path is the local path to an audit policy.
+	Path string
+	// LogDir is the local pth to the directory where logs should be stored.
```
s/pth/path

xiangpengzhao commented on cmd/kubeadm/app/apis/kubeadm/v1alpha1/types.go:
```diff
@@ -240,3 +243,14 @@ type HostPathMount struct {
+type AuditPolicyConfiguration struct {
+	// Path is the local path to an audit policy.
+	Path string `json:"path"`
+	// LogDir is the local pth to the directory where logs should be stored.
```
s/pth/path

xiangpengzhao commented on cmd/kubeadm/app/constants/constants.go:
```diff
@@ -203,6 +203,19 @@ const (
+	AuditPolicyFile = "audit.yaml"
+	// AuditPolicyLogFile is the name of the file audit logs get written to
+	AuditPolicyLogFile = "audit.log"
+	// KubeAuditPolicyLogVolumeName is the name of the volume that will container the audit logs
```
s/container/contain

xiangpengzhao commented on cmd/kubeadm/app/features/features.go:
```diff
@@ -53,6 +56,7 @@ var InitFeatureGates = FeatureList{
 	HighAvailability:     {FeatureSpec: utilfeature.FeatureSpec{Default: false, PreRelease: utilfeature.Alpha}, MinimumVersion: v190, HiddenInHelpText: true},
 	CoreDNS:              {FeatureSpec: utilfeature.FeatureSpec{Default: false, PreRelease: utilfeature.Alpha}, MinimumVersion: v190},
 	DynamicKubeletConfig: {FeatureSpec: utilfeature.FeatureSpec{Default: false, PreRelease: utilfeature.Alpha}, MinimumVersion: v190},
+	Auditing:             {FeatureSpec: utilfeature.FeatureSpec{Default: false, PreRelease: utilfeature.Alpha}},
```
s/Alpha/Beta as per the comment in L45 `// Auditing is beta in 1.8` ?

chuckha replied:
Auditing in k8s is in beta as of 1.8, but the feature to enable it with kubeadm, I assumed, would be considered Alpha

xiangpengzhao replied:
SGTM!

xiangpengzhao commented on cmd/kubeadm/app/util/audit/utils.go:
```diff
@@ -0,0 +1,69 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
s/2017/2018

xiangpengzhao commented on cmd/kubeadm/app/util/audit/utils.go:
```diff
@@ -0,0 +1,69 @@
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	auditv1beta1 "k8s.io/apiserver/pkg/apis/audit/v1beta1"
+
```
extra line

chuckha replied:
weird, maybe my goimports settings are wonky, i'll need to look into this so they are in line with k8s expectations.

xiangpengzhao commented on cmd/kubeadm/app/util/audit/utils.go:
```diff
@@ -0,0 +1,69 @@
+	auditv1beta1 "k8s.io/apiserver/pkg/apis/audit/v1beta1"
+
+	"k8s.io/kubernetes/cmd/kubeadm/app/util"
+
```
same as above

xiangpengzhao commented on cmd/kubeadm/app/util/audit/utils_test.go:
```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
s/2017/2018

xiangpengzhao reviewed (changes requested):
just some nits.

timothysc reviewed (changes requested):
address other comment then lgtm 

/approve

chuckha commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -319,6 +320,21 @@ func (i *Init) Run(out io.Writer) error {
+	if features.Enabled(i.cfg.FeatureGates, features.Auditing) {
+		// Setup the AuditPolicy (either it was passed in and exists or it wasn't passed in and generate a default policy)
+		if i.cfg.AuditPolicyConfiguration.Path != "" {
+			// TODO(chuckha) ensure passed in audit policy is valid so users don't have to find the error in the api server log.
```
I was thinking try to serialize it into a Policy type and see if it was valid, that would probably catch a lot of the cases

timothysc commented:
/test pull-kubernetes-kubemark-e2e-gce

chuckha commented:
/test pull-kubernetes-unit

timothysc reviewed (approved):
/lgtm
